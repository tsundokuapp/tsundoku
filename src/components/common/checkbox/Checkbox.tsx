import { cn } from '@/helpers/twUtils';
import { ICheckbox } from '@/types/Checkbox';

export const Checkbox = ({
  label,
  description,
  name,
  register,
  size = 'sm',
}: ICheckbox) => {
  return (
    <div className="flex">
      <div className="flex h-5 items-center bg-transparent">
        <input
          id="helper-checkbox"
          aria-describedby="helper-checkbox-input"
          type="checkbox"
          className={cn(
            'disabled:border-steel-400 disabled:bg-steel-400 peer relative mt-1 h-4 w-4 shrink-0 appearance-none rounded-sm border-2 border-[#e0e0e0] bg-white checked:border-0 checked:bg-primary focus:outline-none focus:ring-2 focus:ring-primary focus:ring-offset-0 dark:border-slate-700 dark:bg-slate-900 checked:dark:border-0 checked:dark:bg-primary dark:focus:ring-primary',
            {
              'h-4 w-4': size === 'sm',
              'h-5 w-5': size === 'md',
              'h-6 w-6': size === 'lg',
            },
          )}
          {...register(name)}
        />
        <svg
          className={cn(
            'pointer-events-none absolute mt-1 hidden peer-checked:block',
            {
              'h-4 w-4': size === 'sm',
              'h-5 w-5': size === 'md',
              'h-6 w-6': size === 'lg',
            },
          )}
          xmlns="http://www.w3.org/2000/svg"
          viewBox="0 0 24 24"
          fill="none"
          stroke="currentColor"
          stroke-width="4"
          stroke-linecap="round"
          stroke-linejoin="round"
        >
          <polyline points="20 6 9 17 4 12"></polyline>
        </svg>
      </div>
      <div
        className={cn('ms-2 text-sm', {
          'text-md': size === 'md' || 'lg',
        })}
      >
        <label
          htmlFor="helper-checkbox"
          className="font-medium text-gray-900 dark:text-gray-300"
        >
          {label}
        </label>
        {description && (
          <p
            id="helper-checkbox-text"
            className="flex-wrap text-xs font-normal text-gray-500 dark:text-slate-500"
          >
            {description}
          </p>
        )}
      </div>
    </div>
  );
};
