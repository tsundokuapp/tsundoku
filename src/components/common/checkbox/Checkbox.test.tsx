import '@testing-library/jest-dom';
import { fireEvent, render, screen } from '@testing-library/react';
import { type UseFormRegister } from 'react-hook-form';

import { Checkbox } from './Checkbox';

describe('Testes para o componente <Checkbox />', () => {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const fakeRegister: UseFormRegister<any> = jest.fn();

  describe('Testes de renderização', () => {
    it('Deve renderizar Checkbox corretamente', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toBeInTheDocument();
    });

    it('Deve renderizar a label do Checkbox corretamente', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      const label = screen.getByText('Checkbox Teste');

      // Then
      expect(label).toBeInTheDocument();
    });

    it('Deve renderizar o texto de apoio do Checkbox corretamente', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
          description="Texto de apoio do Checkbox Teste"
        />,
      );

      // When
      const description = document.getElementById('helper-checkbox-text');

      // Then
      expect(description).toHaveTextContent('Texto de apoio do Checkbox Teste');
    });

    it('Deve renderizar a label e o texto de apoio do Checkbox simultaneamente', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
          description="Texto de apoio do Checkbox Teste"
        />,
      );

      // When
      const label = screen.getByText('Checkbox Teste');
      const description = document.getElementById('helper-checkbox-text');

      // Then
      expect(label).toHaveTextContent('Checkbox Teste');
      expect(description).toHaveTextContent('Texto de apoio do Checkbox Teste');
    });

    it('Deve renderizar Checkbox com tamanhos diferentes baseado em props', () => {
      // Given
      render(
        <div>
          <Checkbox
            label="Checkbox Teste"
            name="checkboxTeste"
            register={fakeRegister}
            size="lg"
          />
          <Checkbox
            label="Checkbox Teste"
            name="checkboxTeste"
            register={fakeRegister}
            id="checkbox-default"
          />
        </div>,
      );

      // When
      const checkbox = document.getElementById('helper-checkbox');
      const checkboxDefault = document.getElementById('checkbox-default');

      // Then
      expect(checkbox).toHaveClass('h-6 w-6');
      expect(checkboxDefault).toHaveClass('h-4 w-4');
    });
  });

  describe('Testes de interação e comportamento', () => {
    it('Deve iniciar "não-checkado" por padrão', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).not.toBeChecked();
    });

    it('Deve ficar "checkado" quando clicado', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      fireEvent.click(screen.getByRole('checkbox'));
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toBeChecked();
    });

    it('Deve ficar "não-checkado" quando clicado duas vezes', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      fireEvent.click(screen.getByRole('checkbox'));
      const checkboxCheckado = screen.getByRole('checkbox');
      expect(checkboxCheckado).toBeChecked();
      fireEvent.click(screen.getByRole('checkbox'));

      // Then
      expect(screen.getByRole('checkbox')).not.toBeChecked();
    });

    it('Deve tornar Checkbox inacessível quando desabilitado', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
          disabled
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toBeDisabled();
    });
  });

  describe('Testes de acessibilidade', () => {
    it('Deve ter o atributo role igual a checkbox', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toBeInTheDocument();
    });

    it('Deve ter o atributo tabindex igual a 0', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');

      // Then
      expect(checkbox).toHaveAttribute('tabindex', '0');
    });

    it('Dever ser possível focar no checkbox', () => {
      // Given
      render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
          aria-label="Checkbox de teste"
        />,
      );

      // When
      const checkbox = screen.getByRole('checkbox');
      checkbox.focus();

      // Then
      expect(checkbox).toHaveFocus();
    });
  });

  describe('Teste de snapshot', () => {
    it('Deve manter o snapshot do Checkbox', () => {
      // Given
      const { container } = render(
        <Checkbox
          label="Checkbox Teste"
          name="checkboxTeste"
          register={fakeRegister}
        />,
      );

      // Then
      expect(container).toMatchSnapshot();
    });
  });
});
