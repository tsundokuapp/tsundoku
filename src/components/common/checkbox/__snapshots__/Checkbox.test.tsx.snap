// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testes para o componente <Checkbox /> Teste de snapshot Deve manter o snapshot do Checkbox 1`] = `
<div>
  <div
    class="flex"
  >
    <div
      class="flex h-5 items-center bg-transparent"
    >
      <input
        aria-describedby="helper-checkbox-input"
        class="disabled:border-steel-400 disabled:bg-steel-400 checked:bg-primary focus:ring-primary dark:bg-appPrimary checked:dark:bg-primary dark:focus:ring-primary peer relative mt-1 shrink-0 appearance-none rounded-sm border-2 border-[#e0e0e0] bg-white checked:border-0 focus:outline-none focus:ring-2 focus:ring-offset-0 dark:border-slate-700 checked:dark:border-0 h-4 w-4"
        id="helper-checkbox"
        tabindex="0"
        type="checkbox"
      />
      <svg
        class="pointer-events-none absolute mt-1 hidden peer-checked:block h-4 w-4"
        fill="none"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="4"
        viewBox="0 0 24 24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="20 6 9 17 4 12"
        />
      </svg>
    </div>
    <div
      class="ms-2 text-md"
    >
      <label
        class="font-medium text-gray-900 dark:text-gray-300"
        for="helper-checkbox"
      >
        Checkbox Teste
      </label>
    </div>
  </div>
</div>
`;
